### ISSUE TITLE
The shape of the `RandomVariable`s is hard to reason about

### ISSUE THREAD

**Issue by rlouf** [aesara-devs/aesara#1252]
The `RandomVariable`s intialized with different values for `size` have a concrete shape when the `size` is broadcastable, and symbolic otherwise:

```python
import aesara.tensor as at

srng = at.random.randomstream(0)

a_rv = srng.normal(0, 1)
# breakpoint inside `make_node`
# print(size, shape, bcast)
# tensorconstant{[]} tensorconstant{[]} ()
print(a_rv.shape)
# tensorconstant{[]}

a_rv = srng.normal(0, 1, size=(0,))
# breakpoint inside `make_node`
# print(size, shape, bcast)
# tensorconstant{(1,) of 0} tensorconstant{(1,) of 0} (false,)
print(a_rv.shape)
# shape.0


a_rv = srng.normal(0, 1, size=(1,))
# breakpoint inside `make_node`
# print(size, shape, bcast)
# tensorconstant{(1,) of 1} tensorconstant{(1,) of 1} (true,)
print(a_rv.shape)
# tensorconstant{(1,) of 1}


a_rv = srng.normal(0, 1, size=(2,))
# breakpoint inside `make_node`
# print(size, shape, bcast)
# tensorconstant{(1,) of 2} tensorconstant{(1,) of 2} (false,)
print(a_rv.shape)
# shape.0
```

This is problematic for some downstream applications. In `AeHMC`  we create a scalar/vector `inverse_mass_matrix` parameter based on the number of dimensions of the variable, which thus does not have a definite shape. When `a_rv.shape` is a `TensorConstant` the shape checks performed in `Scan` will raise an error.

I would expect all of the above shapes to be `TensorConstant`s, or at the very least to be all symbolic. The shapes are determined when computing the output type with  `TensorType(dtype=dtype, shape=bcast)`.

**Comment by ricardoV94** [comment]
We can output better static shape info beyond broadcastable/not but otherwise I don't se a problem.

You should be able to use `at.specify_shape(input, rv.shape)` if you are having issues with replacing RVs by values or something of that sort.

**Comment by rlouf** [comment]
Turning this into a bug after an offline discussion with @brandonwillard.

**Comment by brandonwillard** [comment]
> Turning this into a bug after an offline discussion with @brandonwillard, `a_rv.shape` should return a `TensorConstant` in every example given in the original comment.

More specifically, it looks like `aesara.tensor.basic.infer_broadcastable` should be returning any and all constant shapes produced by the relatively costly `rewrite_graph` call, instead of discarding that information via the deprecated `broadcastable` encoding.

**Comment by rlouf** [comment]
Indeed, if I insert a breakpoint right before https://github.com/aesara-devs/aesara/blob/739bd49f996974811592bd3384662d9284426b6d/aesara/tensor/basic.py#L1366, `folded_shape` contains the correct type information.

**Comment by brandonwillard** [comment]
> Indeed, if I insert a breakpoint right before
> 
> https://github.com/aesara-devs/aesara/blob/739bd49f996974811592bd3384662d9284426b6d/aesara/tensor/basic.py#L1366
> 
> `folded_shape` contains the correct type information (constant shapes for the cases `size=(0,)` and `size=(2,)` in the original post).

Yeah, I have a fix in #1253, but I should split that off from some of the other experimental stuff in that PR.
